### PR TITLE
Support session EndProfiling() in the CSharp API 

### LINKED ISSUE #3436: Onnxruntime profiler  produces empty log in C# under certain circumstances
**Describe the bug**
Under certain circumstances, C# API produces an empty profiling log.

**Urgency**
Medium

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: 1.2.0
- Python version: 3.7.6
- Visual Studio version (if applicable): VS 2017
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**To Reproduce**
This can be reproduced by instantiating the session as follows:
            options.GraphOptimizationLevel = optLevel;
            options.EnableProfiling = true;
            var session = new InferenceSession(modelPath, options);   

if enclosed in using () then a profiling log file is produced. But if instantiated as above, it produces an empty log file. (0 bytes). 
The bug can be reproduced by editing the Microsoft.ML.OnnxRuntime.PerfTool. See attached file.

**Expected behavior**
Produce a complete log file

**Screenshots**
N/A

**Additional context**
N/A
[Program.txt](https://github.com/microsoft/onnxruntime/files/4440491/Program.txt)

### PR DESCRIPTION
**Pull request by hariharans29**
Fix #3436

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 1]
```diff
@@ -291,6 +291,47 @@ private void CanRunInferenceOnAModel(GraphOptimizationLevel graphOptimizationLev
...
         }
 
+        [Fact]
+        public void InferenceSessionManualDispose()
+        {
+            string modelPath = Path.Combine(Directory.GetCurrentDirectory(), "squeezenet.onnx");
+
+            // Set the graph optimization level for this session.
+            SessionOptions options = new SessionOptions();
+            options.EnableProfiling = false;
```
Comment (skottmckay) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:301]: Why does the unit test for a change to test calling EndProfiling disable profiling?

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 2]
```diff
@@ -687,6 +687,33 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
...
 
+        /// <summary>
+        /// Ends profiling for the session. Returns the profile file name.
+        /// 
+        public string EndProfiling()
+        {
+            IntPtr nameHandle = IntPtr.Zero;
+            string str = null;
+
+            IntPtr status = NativeMethods.OrtSessionEndProfiling(_nativeHandle,
```
Comment (skottmckay) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:698]: Why make this call outside of the 'try'?

Also why no 'catch' with handling or logging of exceptions?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 1 above)
Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs]: I was experimenting and committed this after I marked the PR WIP. I have fixed this.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 2 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:698]: We can make that call outside try. We get back a status which we validate within the try. I followed the same pattern as this - https://github.com/microsoft/onnxruntime/blob/e6da5946d1c7ccaef227f59d9357e74f9d649c1b/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs#L810

File: csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs [hunk 3]
```diff
@@ -127,6 +127,15 @@ public struct OrtApi
...
         public IntPtr ReleaseCustomOpDomain;
+        public IntPtr GetDenotationFromTypeInfo;
+        public IntPtr CastTypeInfoToMapTypeInfo;
+        public IntPtr CastTypeInfoToSequenceTypeInfo;
+        public IntPtr GetMapKeyType;
+        public IntPtr GetMapValueType;
+        public IntPtr GetSequenceElementType;
+        public IntPtr ReleaseMapTypeInfo;
+        public IntPtr ReleaseSequenceTypeInfo;
+        public IntPtr SessionEndProfiling;
```
Comment (hariharans29) [inline csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs:138]: Need to support SessionEndProfiling - so we need to bring in everything until this API to mirror the C struct. The others don't need their corresponding delegates.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 2 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:698]: Also, we don’t really have to encapsulate the call in try as the native API implementation catches any exceptions and returns a status which we handle within the try catch block here.

**Pull request merged**